language-plutus-core/src/Language/PlutusCore/TypeCheck/Internal.hs

```diff
@@ -302,18 +316,62 @@ unfoldFixOf pat arg k = do
             , vArg
             ]
 
--- | Infer the type of a 'Builtin'.
-inferTypeOfBuiltinM
+-- ###########################################
+-- ## Dealing with applications of builtins ##
+-- ###########################################
+
+-- | Infer the type components of a 'Builtin'. The annotation argument
+-- is required for the error if the name isn't found
+inferTypeComponentsOfBuiltinM
     :: (GShow uni, GEq uni, DefaultUni <: uni)
-    => Builtin ann -> TypeCheckM uni ann (Normalized (Type TyName uni ()))
+    => ann -> BuiltinName -> TypeCheckM uni ann (TypeComponents uni)
 -- We have a weird corner case here: the type of a 'BuiltinName' can contain 'TypedBuiltinDyn', i.e.
 -- a static built-in name is allowed to depend on a dynamic built-in type which are not required
 -- to be normalized. For dynamic built-in names we store a map from them to their *normalized types*,
 -- with the normalization happening in this module, but what should we do for static built-in names?
 -- Right now we just renormalize the type of a static built-in name each time we encounter that name.
-inferTypeOfBuiltinM (BuiltinName    _   name) = normalizeType $ typeOfBuiltinName name
+inferTypeComponentsOfBuiltinM ann (StaticBuiltinName name) = getStaticBuiltinTypeComponentsM ann name
 -- TODO: inline this definition once we have only dynamic built-in names.
-inferTypeOfBuiltinM (DynBuiltinName ann name) = lookupDynamicBuiltinNameM ann name
+inferTypeComponentsOfBuiltinM ann (DynBuiltinName name)    = lookupDynamicBuiltinTypeComponentsM ann name
+
+
+-- | Work along the type parameters of a builtin application, building a
+-- mapping from type parameters to the types they're being instantiated at.
+checkBuiltinTypeArgs
+    :: ann
+    -> BuiltinName
+    -> [TyName]                -- type parameters
+    -> [Type TyName uni ann]   -- type instantations
+    -> TypeCheckM uni ann (Norm.TypeMapping TyName uni ())
+checkBuiltinTypeArgs ann bn tyNames tyArgs =
+    case zipExact tyNames tyArgs of
+      Left LeftLonger  -> throwError $ BuiltinTypeErrorE ann (BuiltinUnderInstantiated bn)
+      Left RightLonger -> throwError $ BuiltinTypeErrorE ann (BuiltinOverInstantiated  bn)
+      Right namesAndTypes ->  do
+        mapM_ (mapM_ (\tyarg -> checkKindM (typeAnn tyarg) tyarg (Type ()))) namesAndTypes
+        -- NB: All of the type variables in TypeSchemes have kind *  ^
+        mapping <- mapM (mapM (normalizeType . void)) namesAndTypes  -- ie, normalise all the types in the (TyName, Type) pairs
```

Comment:
Hmm. I'm not so sure about that.  For the final line, I _think_ the best I can do is
```
 for namesAndTypes $ \(tyName,tyArg) -> do
                     tyArg' <- normalizeType . void $ tyArg
                     pure (tyName, tyArg')
```
Maybe just having `mapM` or `traverse` with an explanation would be simpler?  Doing the normalisation before the zipping somehow seems a bit weird as well.  We then do a kind check on the normalised type, and it feels like tht sould be done prior to normalisation.  However, I'll experiment with using extended environments rather than these `TypeMapping` things and see what direction that pushes me in.

Reply:
I think what you wrote is better. I also don't think it's weird to do the normalization before zipping.